File: sdk/node/examples/test_external_task_handler.ts
#!/usr/bin/env npx tsx
/**
 * A3S Code Node.js SDK - External Task Handler Integration Test (TypeScript)
 *
 * Demonstrates the Multi-Machine External Task pattern:
 * 1. Coordinator creates a session with Execute lane set to External mode
 * 2. Agent sends a prompt that triggers bash/write/edit tool calls
 * 3. Coordinator polls pendingExternalTasks() in a background interval
 * 4. Coordinator processes tasks locally (simulating a remote worker)
 * 5. Coordinator completes tasks via completeExternalTask()
 *
 * Run with: npx tsx examples/test_external_task_handler.ts
 */

import { Agent, Session, AgentResult } from '../index.js';
import { execSync } from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as os from 'os';

interface WorkerResult {
  success: boolean;
  output: string;
  exitCode: number;
  error: string | null;
}

interface ExternalTask {
  task_id: string;
  command_type: string;
  payload?: {
    command?: string;
    working_dir?: string;
  };
}

class ExternalTaskHandlerTest {
  private readonly agent: Agent;
  private readonly configPath: string;
  private session: Session | null = null;

  constructor(agent: Agent, configPath: string) {
    this.agent = agent;
    this.configPath = configPath;
  }

  // ── Static Helpers ───────────────────────────────────────────────────────

  static findConfig(): string {
    const homeConfig: string = path.join(os.homedir(), '.a3s', 'config.hcl');
    if (fs.existsSync(homeConfig)) return homeConfig;

    const projectConfig: string = path.resolve(__dirname, '..', '..', '..', '..', '..', '.a3s', 'config.hcl');
    if (fs.existsSync(projectConfig)) return projectConfig;

    throw new Error('Config file not found');
  }

  /**
   * Simulate a remote worker executing a bash command.
   */
  private static workerExecuteBash(command: string, workingDir: string = '.'): WorkerResult {
    try {
      const output: string = execSync(command, {
        cwd: workingDir,
        encoding: 'utf-8',
        timeout: 30000,
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      return { success: true, output, exitCode: 0, error: null };
    } catch (e: unknown) {
      const err = e as { stdout?: string; status?: number; stderr?: string; message?: string };
      return {
        success: false,
        output: err.stdout || '',
        exitCode: err.status || 1,
        error: err.stderr || err.message || 'Unknown error',
      };
    }
  }

  // ── Instance Test Methods ────────────────────────────────────────────────

  async testExternalTaskHandler(): Promise<void> {
    console.log('\n📦 Test 1: External Task Handler (Execute Lane → External)');
    console.log('-'.repeat(80));

    // 1. Create session with queue enabled
    const session: Session = this.agent.session('.', {
      queueConfig: {
        enableAllFeatures: true,
        timeoutMs: 60000,
      },
    });

    // 2. Route Execute lane to External mode
    await session.setLaneHandler('execute', {
      mode: 'external',
      timeoutMs: 60000,
    });

    console.log('✓ Session created with Execute lane → External mode');
    console.log('  Query lane: Internal (read, glob, grep run locally)');
    console.log('  Execute lane: External (bash, write, edit → ExternalTask)');
    console.log();

    // 3. Start a background poller for external tasks
    const start: number = Date.now();
    let externalTasksProcessed: number = 0;
    let stopPolling: boolean = false;

    const pollInterval: ReturnType<typeof setInterval> = setInterval(async (): Promise<void> => {
      if (stopPolling) return;

      try {
        const tasks = await session.pendingExternalTasks() as unknown as ExternalTask[];
        if (!Array.isArray(tasks)) return;

        for (const task of tasks) {
          const taskId: string = task.task_id;
          const cmdType: string = task.command_type;
          console.log(`  📥 ExternalTaskPending: ${taskId.slice(0, 8)} (${cmdType})`);

          // Execute the task (simulating a remote worker)
          let result: WorkerResult;
          if (cmdType === 'bash') {
            const cmd: string = task.payload?.command || "echo 'no command'";
            const cwd: string = task.payload?.working_dir || '.';
            result = ExternalTaskHandlerTest.workerExecuteBash(cmd, cwd);
          } else {
            result = {
              success: true,
              output: `External handler processed: ${cmdType}`,
              exitCode: 0,
              error: null,
            };
          }

          console.log(`  🔧 Worker result: success=${result.success}, exit_code=${result.exitCode}`);
          const preview: string = (result.output || '').trim().slice(0, 60);
          if (preview) console.log(`     Output: ${preview}`);

          // Complete the external task
          const completed: boolean = await session.completeExternalTask(taskId, {
            success: result.success,
            result: { output: result.output, exit_code: result.exitCode },
            error: result.error,
          });

          if (completed) {
            externalTasksProcessed++;
            console.log(`  📤 Task ${taskId.slice(0, 8)} completed and returned to agent`);
          }
        }
      } catch (e: unknown) {
        // Session might be done
        const err = e as { message?: string };
        if (!err.message?.includes('exhausted')) {
          console.log(`  ⚠ Poll error: ${err.message}`);
        }
      }
    }, 200); // Poll every 200ms

    // 4. Send the prompt — agent will produce ExternalTask objects for bash calls
    //    Note: Node SDK stream() collects all events, so we use send() + polling
    let result: AgentResult;
    try {
      result = await session.send(
        "Run these bash commands and tell me the results:\n" +
        "1. echo 'Hello from external worker'\n" +
        "2. date '+%Y-%m-%d %H:%M:%S'\n" +
        "3. uname -s"
      );
    } finally {
      stopPolling = true;
      clearInterval(pollInterval);
    }

    const duration: number = (Date.now() - start) / 1000;

    console.log();
    console.log('-'.repeat(80));
    console.log('📊 Results:');
    console.log(`  Duration: ${duration.toFixed(2)}s`);
    console.log(`  External tasks processed: ${externalTasksProcessed}`);
    console.log(`  Response length: ${result.text.length} chars`);
    console.log(`  Tool calls: ${result.toolCallsCount}`);

    // Store session for subsequent tests
    this.session = session;
  }

  async testHybridMode(): Promise<void> {
    console.log('\n\n📦 Test 2: Hybrid Mode (Execute Lane → Hybrid)');
    console.log('-'.repeat(80));

    if (!this.session) throw new Error('testExternalTaskHandler must run first');

    // Switch to Hybrid mode
    await this.session.setLaneHandler('execute', {
      mode: 'hybrid',
      timeoutMs: 60000,
    });

    console.log('✓ Execute lane switched to Hybrid mode');
    console.log('  Tools execute locally AND emit ExternalTaskPending events');
    console.log();

    const start: number = Date.now();
    const result: AgentResult = await this.session.send("Run: echo 'hybrid mode test'");
    const duration: number = (Date.now() - start) / 1000;

    console.log(`✓ Completed in ${duration.toFixed(2)}s`);
    console.log(`  Response: ${result.text.slice(0, 120)}`);
    console.log(`  Tool calls: ${result.toolCallsCount}`);
  }

  async testDynamicLaneSwitching(): Promise<void> {
    console.log('\n\n📦 Test 3: Dynamic Lane Switching');
    console.log('-'.repeat(80));

    if (!this.session) throw new Error('testExternalTaskHandler must run first');

    // Switch back to Internal
    await this.session.setLaneHandler('execute', {
      mode: 'internal',
      timeoutMs: 60000,
    });

    console.log('✓ Execute lane switched back to Internal mode');

    const start: number = Date.now();
    const result: AgentResult = await this.session.send("Run: echo 'back to internal mode'");
    const duration: number = (Date.now() - start) / 1000;

    console.log(`✓ Completed in ${duration.toFixed(2)}s`);
    console.log(`  Response: ${result.text.slice(0, 120)}`);
  }

  // ── Run All ───────────────────────────────────────────────────────────────

  async runAll(): Promise<void> {
    console.log('🚀 A3S Code - External Task Handler Integration Test\n');
    console.log('='.repeat(80));
    console.log(`📄 Config: ${this.configPath}`);
    console.log('='.repeat(80));

    await this.testExternalTaskHandler();
    await this.testHybridMode();
    await this.testDynamicLaneSwitching();

    console.log('\n' + '='.repeat(80));
    console.log('✅ All external task handler tests completed!');
    console.log('='.repeat(80));
  }
}

async function main(): Promise<void> {
  const configPath = ExternalTaskHandlerTest.findConfig();
  const agent = await Agent.create(configPath);
  const test = new ExternalTaskHandlerTest(agent, configPath);
  await test.runAll();
}

main().catch(console.error);

File: sdk/python/examples/test_external_task_handler.py
#!/usr/bin/env python3
"""
A3S Code Python SDK - External Task Handler Integration Test

Demonstrates the Multi-Machine External Task pattern:
1. Coordinator creates a session with Execute lane set to External mode
2. Agent streams a prompt that triggers bash/write/edit tool calls
3. ExternalTaskPending events fire — coordinator polls pending tasks
4. Coordinator processes tasks locally (simulating a remote worker)
5. Coordinator completes tasks via complete_external_task()

Run with: python examples/test_external_task_handler.py
"""

import subprocess
import time
import threading
from pathlib import Path
from a3s_code import Agent, SessionQueueConfig


class ExternalTaskHandlerTest:
    """Integration test for the external task handler pattern."""

    def __init__(self, agent: Agent, config_path: str) -> None:
        self.agent = agent
        self.config_path = config_path
        self._session = None

    @staticmethod
    def find_config_path() -> str:
        """Find config file."""
        config_path = Path.home() / ".a3s" / "config.hcl"
        if config_path.exists():
            return str(config_path)

        project_config = Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "config.hcl"
        if project_config.exists():
            return str(project_config)

        raise FileNotFoundError("Config file not found")

    @staticmethod
    def worker_execute_bash(command: str, working_dir: str = ".") -> dict:
        """Simulate a remote worker executing a bash command."""
        try:
            result = subprocess.run(
                ["sh", "-c", command],
                cwd=working_dir,
                capture_output=True,
                text=True,
                timeout=30,
            )
            return {
                "success": result.returncode == 0,
                "output": result.stdout,
                "exit_code": result.returncode,
                "error": result.stderr if result.returncode != 0 else None,
            }
        except Exception as e:
            return {
                "success": False,
                "output": "",
                "exit_code": 1,
                "error": str(e),
            }

    def test_external_task_handler(self) -> None:
        """Test 1: External mode — Execute lane tasks routed to external handler."""
        print("\n📦 Test 1: External Task Handler (Execute Lane → External)")
        print("-" * 80)

        # 1. Create session with queue enabled
        qc = SessionQueueConfig()
        qc.with_lane_features()
        qc.set_timeout(60000)
        self._session = self.agent.session(".", queue_config=qc)

        # 2. Route Execute lane to External mode
        self._session.set_lane_handler("execute", mode="external", timeout_ms=60000)

        print("✓ Session created with Execute lane → External mode")
        print("  Query lane: Internal (read, glob, grep run locally)")
        print("  Execute lane: External (bash, write, edit → ExternalTask)")
        print()

        # 3. Stream a prompt that will trigger Execute-lane tools (bash)
        start = time.time()
        text_output = ""

        # Start a background thread to poll and process external tasks
        stop_polling = threading.Event()
        poll_results = {"count": 0}

        def poll_external_tasks() -> None:
            """Background poller that checks for pending external tasks."""
            while not stop_polling.is_set():
                try:
                    tasks = self._session.pending_external_tasks()
                    for task in tasks:
                        task_id = task["task_id"]
                        cmd_type = task["command_type"]
                        print(f"  📥 ExternalTaskPending: {task_id[:8]} ({cmd_type})")

                        # Execute the task (simulating a remote worker)
                        if cmd_type == "bash":
                            cmd = task.get("payload", {}).get("command", "echo 'no command'")
                            cwd = task.get("payload", {}).get("working_dir", ".")
                            result = self.worker_execute_bash(cmd, cwd)
                        else:
                            result = {
                                "success": True,
                                "output": f"External handler processed: {cmd_type}",
                                "exit_code": 0,
                                "error": None,
                            }

                        print(f"  🔧 Worker result: success={result['success']}, exit_code={result['exit_code']}")
                        output_preview = (result["output"] or "").strip()[:60]
                        if output_preview:
                            print(f"     Output: {output_preview}")

                        # Complete the external task
                        completed = self._session.complete_external_task(
                            task_id,
                            success=result["success"],
                            result={"output": result["output"], "exit_code": result["exit_code"]},
                            error=result["error"],
                        )

                        if completed:
                            poll_results["count"] += 1
                            print(f"  📤 Task {task_id[:8]} completed and returned to agent")
                except Exception as e:
                    # Session might be done
                    if "exhausted" not in str(e).lower():
                        print(f"  ⚠ Poll error: {e}")

                time.sleep(0.2)  # Poll every 200ms

        # Start the poller
        poller = threading.Thread(target=poll_external_tasks, daemon=True)
        poller.start()

        # 4. Stream the prompt — events arrive as the agent works
        for event in self._session.stream(
            "Run these bash commands and tell me the results:\n"
            "1. echo 'Hello from external worker'\n"
            "2. date '+%Y-%m-%d %H:%M:%S'\n"
            "3. uname -s"
        ):
            if event.event_type == "text_delta":
                text_output += event.text or ""
                print(event.text or "", end="", flush=True)
            elif event.event_type == "tool_start":
                print(f"\n  🔨 Tool: {event.tool_name}")
            elif event.event_type == "end":
                print()
                break
            elif event.event_type == "error":
                print(f"\n  ❌ Error: {event.error}")
                break

        # Stop the poller
        stop_polling.set()
        poller.join(timeout=2)

        duration = time.time() - start
        external_tasks_processed = poll_results["count"]

        print()
        print("-" * 80)
        print("📊 Results:")
        print(f"  Duration: {duration:.2f}s")
        print(f"  External tasks processed: {external_tasks_processed}")
        print(f"  Response length: {len(text_output)} chars")

    def test_hybrid_mode(self) -> None:
        """Test 2: Hybrid mode — local execution + external notification."""
        print("\n\n📦 Test 2: Hybrid Mode (Execute Lane → Hybrid)")
        print("-" * 80)

        # Switch to Hybrid mode
        self._session.set_lane_handler("execute", mode="hybrid", timeout_ms=60000)

        print("✓ Execute lane switched to Hybrid mode")
        print("  Tools execute locally AND emit ExternalTaskPending events")
        print()

        start = time.time()
        result = self._session.send("Run: echo 'hybrid mode test'")
        duration = time.time() - start

        print(f"✓ Completed in {duration:.2f}s")
        print(f"  Response: {result.text[:120]}")
        print(f"  Tool calls: {result.tool_calls_count}")

    def test_dynamic_lane_switching(self) -> None:
        """Test 3: Dynamic lane switching."""
        print("\n\n📦 Test 3: Dynamic Lane Switching")
        print("-" * 80)

        # Switch back to Internal
        self._session.set_lane_handler("execute", mode="internal", timeout_ms=60000)

        print("✓ Execute lane switched back to Internal mode")

        start = time.time()
        result = self._session.send("Run: echo 'back to internal mode'")
        duration = time.time() - start

        print(f"✓ Completed in {duration:.2f}s")
        print(f"  Response: {result.text[:120]}")

    def run_all(self) -> None:
        """Run all external task handler tests."""
        print("🚀 A3S Code - External Task Handler Integration Test\n")
        print("=" * 80)
        print(f"📄 Config: {self.config_path}")
        print("=" * 80)

        self.test_external_task_handler()
        self.test_hybrid_mode()
        self.test_dynamic_lane_switching()

        print("\n" + "=" * 80)
        print("✅ All external task handler tests completed!")
        print("=" * 80)


if __name__ == "__main__":
    config_path = ExternalTaskHandlerTest.find_config_path()
    agent = Agent.create(config_path)
    test = ExternalTaskHandlerTest(agent, config_path)
    test.run_all()
